### ISSUE TITLE
When using client Logger, connections are exhausted in recursive calls to .expect

### ISSUE THREAD

**Issue by rhyskeepence** [http4s/http4s#1833]
I'm integrating with an api, which appears to have exposed an issue (or misunderstanding) with how .expect frees connections.

The api is a asynchronous service - we post a request, poll until a status endpoint returns 200 and a particular value in the response to indicate completion, and then fetch the result from a third endpoint. Frankly I would have designed it differently, but that's how it is. 

I'm doing the retry of status calls as a recursive call. The issue is when the status endpoint returns non-2xx. In this case, connections are not being disposed, and after 10 retries (the default number of client connections) the program hangs.

The following gist reproduces the issue with a stub server: https://gist.github.com/rhyskeepence/b575766c197f179e71c28e767ad490d9

**Update:** the problem is only exhibited when using the 0.18.10 client Logger. The behavior is correct when removing the logger (connections are disposed of). I also noticed that the response is not logged when the status is non-2xx.

Thanks

**Comment by rossabaker** [comment]
Thanks.  This is something we can work with!

I can reproduce the hang, but I'm seeing the non-2xx responses.  Does your log not look like this?

```
[info] 08:47:32.522 [scala-execution-context-global-34] INFO  o.h.client.middleware.RequestLogger - HTTP/1.1 GET http://localhost:8080/status Headers(Accept: text/*) 
[info] 08:47:32.523 [blaze-nio1-acceptor] INFO  o.h.blaze.channel.ServerChannelGroup - Connection to /127.0.0.1:42712 accepted at Wed May 09 08:47:32 EDT 2018.
[info] 08:47:32.527 [scala-execution-context-global-31] INFO  o.h.client.middleware.ResponseLogger - HTTP/1.1 404 Not Found Headers(Content-Type: text/plain; charset=UTF-8, Date: Wed, 09 May 2018 12:47:32 GMT, Content-Length: 17) 
```

**Comment by rossabaker** [comment]
Okay, I'll look into that on a separate PR.  That sounds like it might be the problem with the loggers when the body is not fully consumed.

**Comment by rossabaker** [comment]
The deadlock is fixed and the logging improved in 0.18.11.  Note the caveat on #1846.
